## Dataframe explore

In [31]:
import pandas as pd 
import numpy as np 
df =pd.read_csv("data/messy_retail_sales_practice.csv")


# Inspection 

#### DataFrame head() explore 
- This method primarily use for inspection and display first 5 rows plus included header
- df.head(n=5), here n is number of rows should display , by default its **5** 

In [32]:
df.head()

,OrderID,OrderDate,CustomerID,CustomerName,CustomerAge,Email,Region,City,Category,Product,Quantity,UnitPrice,DiscountPct,PaymentMethod,SalesRep,Returned,SatisfactionScore,DeliveryDays,Notes
0,ORD-10136,2025-09-27,C0083,Rahim Miah,58.0,rahim.miah36@example.com,West,Pabna,Electronics,Wireless Mouse,2,18.41,0.0,Card,Priya,No,1.0,9.0,First-time customer
1,ORD-10237,2026-05-21,C0156,Rafi Islam,68.0,rafi.islam54@example.com,South,Khulna,Home,Storage Box,1,11.62,0.0,Cash,Shakib,No,3.0,3.0,First-time customer
2,ORD-10083,2025-07-30,C0110,Mim Ahmed,63.0,mim.ahmed28@example.com,Central,Dhaka,Clothing,Sneakers,2,48.56,5.0,Cash,Priya,No,5.0,3.0,First-time customer
3,ORD-10299,2025-01-23,C0018,Rafi Chowdhury,51.0,rafi.chowdhury65@example.com,East,Cumilla,Electronics,Headphones,1,46.77,0.0,Cash,Shakib,Yes,1.0,4.0,Address confirmed
4,ORD-10424,01/07/2025,C0025,Mahmud Rahman,56.0,mahmud.rahman48@example.com,South,Jessore,Home,Water Bottle,2,10.69,0.0,Cash,Rashed,No,4.0,1.0,Address confirmed


#### DataFrame info() explore
> shows a concise summary of the DataFrame.
**It displays:**
- Number of rows
- Column names
- Non-null values in each column
- Data type of each column
- Memory usage

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 460 entries, 0 to 459
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   OrderID            460 non-null    object 
 1   OrderDate          460 non-null    object 
 2   CustomerID         460 non-null    object 
 3   CustomerName       438 non-null    object 
 4   CustomerAge        451 non-null    float64
 5   Email              459 non-null    object 
 6   Region             460 non-null    object 
 7   City               450 non-null    object 
 8   Category           460 non-null    object 
 9   Product            460 non-null    object 
 10  Quantity           460 non-null    int64  
 11  UnitPrice          447 non-null    float64
 12  DiscountPct        443 non-null    float64
 13  PaymentMethod      460 non-null    object 
 14  SalesRep           460 non-null    object 
 15  Returned           460 non-null    object 
 16  SatisfactionScore  410 non

#### DataFrame describe()  explore
> `df.describe()` shows summary statistics for numerical columns.
**It displays:**

- Count
- Mean
- Standard deviation
- Minimum
- 25th percentile
- Median
- 75th percentile
- Maximum

In [34]:
df.describe()

,CustomerAge,Quantity,UnitPrice,DiscountPct,SatisfactionScore,DeliveryDays
count,451.000000,460.000000,447.000000,443.000000,410.000000,449.000000
mean,43.611973,2.063043,45.542394,6.094808,2.997561,5.741648
std,14.976507,1.200192,472.146356,7.474545,1.444150,3.377985
min,18.000000,-2.000000,3.460000,0.000000,1.000000,1.000000
25%,30.000000,1.000000,11.880000,0.000000,2.000000,3.000000
50%,42.000000,2.000000,18.510000,5.000000,3.000000,6.000000
75%,57.000000,3.000000,29.110000,10.000000,4.000000,8.000000
max,70.000000,5.000000,9999.990000,85.000000,9.000000,45.000000


#### DataFrame shape explore
> `df.shape` returns the number of rows and columns in the DataFrame.
**Example output:**
```python
(460, 19)
```
This means:
- 460 rows
- 19 columns

In [35]:
df.shape

(460, 19)

## DataFrame Cleaning 

#### DataFrame dropna() explore
**df.dropna() removes rows containing missing values.**

```python
df.dropna()
```

**To permanently update the DataFrame:**
```python
df.dropna(inplace=True)
```
**To remove rows only when specific columns contain missing values:**
```python
df.dropna(subset=["CustomerName", "Email"])
```

In [36]:
dropnaDf= pd.DataFrame({
    "Name": ["Sara", "Rahim", np.nan, "Mim"],
    "Age": [25, np.nan, np.nan, 22],
    "City": ["Dhaka", "Khulna", np.nan, np.nan]
})
print('Clean DataFrame before executing any dropna')
display(dropnaDf)

copy_df =dropnaDf.copy()
# It will remove entire rows if any column contain NaN
copy_df=copy_df.dropna()
print('After applying dropna whiout any parameter')
display(copy_df)

# Drop if any rows contain all columns as NaN , 
#If any one columns contain non NaN , It will not replace it
copy_df=dropnaDf.copy()
copy_df=copy_df.dropna(how='all')
print('After  applying dropna (how=\'all\')')
display(copy_df)

# Filter specific columns , If it containes any NaN value , remove specific raws 
# Not impact other columns even other columns contain NaN 
copy_df =dropnaDf.copy()
copy_df=copy_df.dropna(subset=["Age"])
print("After applying dropna(subset=[\"Age\"])\n\n")
display(copy_df.reset_index())

Clean DataFrame before executing any dropna


,Name,Age,City
0,Sara,25.0,Dhaka
1,Rahim,NaN,Khulna
2,NaN,NaN,NaN
3,Mim,22.0,NaN


After applying dropna whiout any parameter


,Name,Age,City
0,Sara,25.0,Dhaka


After  applying dropna (how='all')


,Name,Age,City
0,Sara,25.0,Dhaka
1,Rahim,NaN,Khulna
3,Mim,22.0,NaN


After applying dropna(subset=["Age"])




,index,Name,Age,City
0,0,Sara,25.0,Dhaka
1,3,Mim,22.0,NaN


#### DataFrame fillna() explore
**df.fillna() replace rows containing missing values, with given value.**

```python
df.fillna()
```

**To update specific columns to the DataFrame:**
```python
df["City"] = df["City"].fillna("Unknown")
```
**To limit the fillna(value,limit=1)**
```python
df["Age"].fillna(0, limit=1)
```

In [38]:
fillna_df = pd.DataFrame({
    "Name": ["Sara", "Rahim", np.nan, "Mim"],
    "Age": [25, np.nan, 30, np.nan],
    "City": ["Dhaka", "Khulna", np.nan, np.nan],
    "Score": [80, 70, np.nan, 90]
})
print("Original DataFrame without any change")
display(fillna_df)

# After applying fillna() 
copy_df=fillna_df.copy()
copy_df=copy_df.fillna("Unknown")
print("\nAfter applying fillna(\"Unknown\") , it update every point in each rows which column containg NaN value:")
display(copy_df)


copy_df =fillna_df.copy()
print("\nAfter applying df[\'City\'].fillna(\"Rajshahi\")")
copy_df["City"]=copy_df["City"].fillna("Rajshahi")
display(copy_df)


#df["Age"].fillna(0, limit=1)
copy_df =fillna_df.copy()
copy_df["Age"]=copy_df["Age"].fillna(20,limit=1)
print("\nAfter applying fillna(20,limit=1) I should see one First Age NaN will change")
display(copy_df)


Original DataFrame without any change


,Name,Age,City,Score
0,Sara,25.0,Dhaka,80.0
1,Rahim,NaN,Khulna,70.0
2,NaN,30.0,NaN,NaN
3,Mim,NaN,NaN,90.0



After applying fillna("Unknown") , it update every point in each rows which column containg NaN value:


,Name,Age,City,Score
0,Sara,25.0,Dhaka,80.0
1,Rahim,Unknown,Khulna,70.0
2,Unknown,30.0,Unknown,Unknown
3,Mim,Unknown,Unknown,90.0



After applying df['City'].fillna("Rajshahi")


,Name,Age,City,Score
0,Sara,25.0,Dhaka,80.0
1,Rahim,NaN,Khulna,70.0
2,NaN,30.0,Rajshahi,NaN
3,Mim,NaN,Rajshahi,90.0



After applying fillna(20,limit=1) I should see one First Age NaN will change


,Name,Age,City,Score
0,Sara,25.0,Dhaka,80.0
1,Rahim,20.0,Khulna,70.0
2,NaN,30.0,NaN,NaN
3,Mim,NaN,NaN,90.0


#### DataFrame drop() explore
**df.fillna() replace rows containing missing values, with given value.**

```python
df.drop()
```

**To update specific columns to the DataFrame:**
```python
df["City"] = df["City"].fillna("Unknown")
```
**To limit the fillna(value,limit=1)**
```python
df["Age"].fillna(0, limit=1)
```

In [39]:

drop_df = pd.DataFrame({
    "Name": ["Sara", "Rahim", "Mim", "Hasan"],
    "Age": [25, 30, 22, 35],
    "City": ["Dhaka", "Khulna", "Rajshahi", "Sylhet"]
})

print("Original DataFrame")
display(drop_df)

#Drop by index
copy_df=drop_df.copy()
copy_df=copy_df.drop(1)
print("\nAfter apply drop(1) first indexed value will erase")
display(copy_df)


copy_df=drop_df.copy()
print("\nAfter applying drop([1,3]) index 1 and 3 will be erase")
copy_df=copy_df.drop([1,3])
display(copy_df)
print("After resetting index")
copy_df=copy_df.reset_index()
display(copy_df)

#Drop a specific column drop(columns=["Age"])
copy_df=drop_df.copy()
copy_df=copy_df.drop(columns=["Age"])
#alise copy_df=copy_df.drop("Age",axis=1)
print("\nAfter deleting columns \"Age\" ")
display(copy_df)

#Drop raw based condition df.drop(df[df["Age"] < 25].index)
copy_df =drop_df.copy()
print("\nAfter applying df.drop(df[df[\"Age\"] < 25].index) , age less than 25 will be erase")
copy_df=copy_df.drop(copy_df[copy_df["Age"] < 25].index)
display(copy_df)


Original DataFrame


,Name,Age,City
0,Sara,25,Dhaka
1,Rahim,30,Khulna
2,Mim,22,Rajshahi
3,Hasan,35,Sylhet



After apply drop(1) first indexed value will erase


,Name,Age,City
0,Sara,25,Dhaka
2,Mim,22,Rajshahi
3,Hasan,35,Sylhet



After applying drop([1,3]) index 1 and 3 will be erase


,Name,Age,City
0,Sara,25,Dhaka
2,Mim,22,Rajshahi


After resetting index


,index,Name,Age,City
0,0,Sara,25,Dhaka
1,2,Mim,22,Rajshahi



After deleting columns "Age" 


,Name,City
0,Sara,Dhaka
1,Rahim,Khulna
2,Mim,Rajshahi
3,Hasan,Sylhet



After applying df.drop(df[df["Age"] < 25].index) , age less than 25 will be erase


,Name,Age,City
0,Sara,25,Dhaka
1,Rahim,30,Khulna
3,Hasan,35,Sylhet


#### DataFrame rename() explore
**df.rename() columns name and idexes values as well**

```python
df.rename(columns={"old_name":"new_name")
```

**To update index to the DataFrame:**
```python
df =df.rename(index={
    0:'student_sarah',
    1:'student_alex',
    2:'student_john'
})
```
**To limit the fillna(value,limit=1)**
```python
df.rename(
    mapper:'Renamer|None'=None,
    *, 
    index : 'Renamer|None'=None,
    columns:'Renamer|None'=None,
    inplace :'bool'=False
    ) 
```

In [40]:
rename_df = pd.DataFrame({
    "name": ["Sara", "Alex", "John"],
    "age": [25, 30, 28],
    "city": ["Dhaka", "London", "New York"]
})

print("Original DataFrame")
display(rename_df)

#Rname all columns 
copy_df=rename_df.copy()
modify_columns={
    "name":"user_name",
    "age":"user_age",
    "city":"user_city"
}
copy_df=copy_df.rename(columns=modify_columns)
print("\nAfter modifying columns name")
display(copy_df)

copy_df=copy_df.rename(index={
    0:'student_sarah',
    1:'student_alex',
    2:'student_john'}
                      )
print("\nAter rename indexes")
display(copy_df)

Original DataFrame


,name,age,city
0,Sara,25,Dhaka
1,Alex,30,London
2,John,28,New York



After modifying columns name


,user_name,user_age,user_city
0,Sara,25,Dhaka
1,Alex,30,London
2,John,28,New York



Ater rename indexes


,user_name,user_age,user_city
student_sarah,Sara,25,Dhaka
student_alex,Alex,30,London
student_john,John,28,New York


### DataFrame astype() explore
**df.astype() type conversion the columns value type**

```python
converted_df = df.astype({
    "age": "int64",
    "salary": "float64",
    "is_active": "bool"
})
```


In [41]:


astype_df = pd.DataFrame({
    "age": ["25", "30", "28"],
    "salary": ["25000.50", "32000.75", "28000.00"],
    "is_active": [1, 0, 1]
})

print("Original DataFrame:")
print(astype_df)

print("\nOriginal data types:")
print(astype_df.dtypes)

# Convert multiple columns to different data types
converted_df = astype_df.astype({
    "age": "int64",
    "salary": "float64",
    "is_active": "bool"
})

print("\nConverted DataFrame:")
print(converted_df)

print("\nConverted data types:")
print(converted_df.dtypes)

Original DataFrame:
  age    salary  is_active
0  25  25000.50          1
1  30  32000.75          0
2  28  28000.00          1

Original data types:
age          object
salary       object
is_active     int64
dtype: object

Converted DataFrame:
   age    salary  is_active
0   25  25000.50       True
1   30  32000.75      False
2   28  28000.00       True

Converted data types:
age            int64
salary       float64
is_active       bool
dtype: object


# Data Selection and Filtering 

### DataFrame `DataFrame.loc[]` explore

`DataFrame.loc[]` selects or updates rows and columns using labels or conditions.

#### Example
> Select one rows
```python
print(df.loc[1])
```
> Select rows and columns
```python
print(df.loc[0:1, ["name", "age"]])
```
> Select row using condition
```python

print(df.loc[df["city"] == "Dhaka"])
```
> Update values on condition
```python
df.loc[df["city"] == "Dhaka", "city"] = "Bangladesh"
```
>Update one cell by index and column name
```python
df.loc[0, "age"] = 26

```

#### Basic syntax

```python
df.loc[row_selector, column_selector]
```

In [42]:

loc_df = pd.DataFrame({
    "name": ["Sara", "Alex", "John", "Emma"],
    "age": [25, 30, 22, 28],
    "city": ["Dhaka", "London", "Dhaka", "Paris"]
})

print("Original DataFrame:")
print(loc_df)

# Select one row by index label
row = loc_df.loc[1]

print("\nRow at index 1:")
print(row)

# Select specific rows and columns
selected_data = loc_df.loc[1:2, ["name", "age"]]

print("\nSelected rows and columns:")
print(selected_data)

# Select rows using a condition
dhaka_people = loc_df.loc[loc_df["city"] == "Dhaka"]

print("\nPeople from Dhaka:")
print(dhaka_people)

# Select specific columns using a condition
adult_names = loc_df.loc[loc_df["age"] >= 25, ["name", "age"]]

print("\nPeople aged 25 or above:")
print(adult_names)

# Update values using loc
loc_df.loc[loc_df["city"] == "Dhaka", "city"] = "Bangladesh"

print("\nUpdated DataFrame:")
print(loc_df)

Original DataFrame:
   name  age    city
0  Sara   25   Dhaka
1  Alex   30  London
2  John   22   Dhaka
3  Emma   28   Paris

Row at index 1:
name      Alex
age         30
city    London
Name: 1, dtype: object

Selected rows and columns:
   name  age
1  Alex   30
2  John   22

People from Dhaka:
   name  age   city
0  Sara   25  Dhaka
2  John   22  Dhaka

People aged 25 or above:
   name  age
0  Sara   25
1  Alex   30
3  Emma   28

Updated DataFrame:
   name  age        city
0  Sara   25  Bangladesh
1  Alex   30      London
2  John   22  Bangladesh
3  Emma   28       Paris


### Pandas `DataFrame.iloc[]`

>`DataFrame.iloc[]` selects or updates rows and columns using integer positions.

## Example
> Select first row
```python
print(df.iloc[0])
```
>Select rowws from position 0 to 1
```python
print(df.iloc[0:2])
```
> Select specific rows and columns 
```python
print(df.iloc[0:2, 0:2])
```
> Select one specific value
```python
print(df.iloc[1, 2])
```
>Update one cell which is first rows 2nd column [0,1]
```python
df.iloc[0, 1] = 26
```
> Update an entire column
```python
df.iloc[:, 2] = ["Bangladesh", "UK", "France"]

```

#### Basic syntax

```python
df.iloc[row_position, column_position]
```

#### Common examples

```python
df.iloc[0]          # First row
df.iloc[-1]         # Last row
df.iloc[:, 0]       # First column
df.iloc[0:3]        # First three rows
df.iloc[0:3, 0:2]   # First three rows and first two columns
```


In [43]:

iloc_df = pd.DataFrame({
    "name": ["Sara", "Alex", "John", "Emma", "David"],
    "age": [25, 30, 22, 28, 35],
    "city": ["Dhaka", "London", "Paris", "Tokyo", "Sydney"],
    "score": [85, 90, 78, 88, 92]
})

print("Original DataFrame:")
print(iloc_df)

# Select the first row
first_row = iloc_df.iloc[0]
print("\nFirst row:")
print(first_row)

# Select the last row
last_row = iloc_df.iloc[-1]
print("\nLast row:")
print(last_row)

# Select the first three rows
first_three_rows = iloc_df.iloc[0:3]
print("\nFirst three rows:")
print(first_three_rows)

# Select the first column
first_column = iloc_df.iloc[:, 0]
print("\nFirst column:")
print(first_column)

# Select multiple columns
selected_columns = iloc_df.iloc[:, [0, 2]]
print("\nSelected columns:")
print(selected_columns)

# Select specific rows and columns
selected_data = iloc_df.iloc[1:4, 0:3]
print("\nSelected rows and columns:")
print(selected_data)

# Select one specific value
selected_value = iloc_df.iloc[1, 2]
print("\nSpecific value:")
print(selected_value)

# Select non-consecutive rows
selected_rows = iloc_df.iloc[[0, 2, 4]]
print("\nNon-consecutive rows:")
print(selected_rows)

# Select non-consecutive rows and columns
selected_rows_columns = iloc_df.iloc[[0, 2, 4], [0, 3]]
print("\nNon-consecutive rows and columns:")
print(selected_rows_columns)

# Update one cell
iloc_df.iloc[0, 1] = 26

# Update multiple cells in one column
iloc_df.iloc[0:3, 3] = [87, 93, 80]

# Update an entire column
iloc_df.iloc[:, 2] = [
    "Bangladesh",
    "United Kingdom",
    "France",
    "Japan",
    "Australia"
]

print("\nUpdated DataFrame:")
print(iloc_df)

# Select all rows except the last row
all_except_last = iloc_df.iloc[:-1]
print("\nAll rows except the last row:")
print(all_except_last)

# Select all columns except the last column
all_except_last_column = iloc_df.iloc[:, :-1]
print("\nAll columns except the last column:")
print(all_except_last_column)

# Select the last two rows and last two columns
last_two_rows_columns = iloc_df.iloc[-2:, -2:]
print("\nLast two rows and last two columns:")
print(last_two_rows_columns)

Original DataFrame:
    name  age    city  score
0   Sara   25   Dhaka     85
1   Alex   30  London     90
2   John   22   Paris     78
3   Emma   28   Tokyo     88
4  David   35  Sydney     92

First row:
name      Sara
age         25
city     Dhaka
score       85
Name: 0, dtype: object

Last row:
name      David
age          35
city     Sydney
score        92
Name: 4, dtype: object

First three rows:
   name  age    city  score
0  Sara   25   Dhaka     85
1  Alex   30  London     90
2  John   22   Paris     78

First column:
0     Sara
1     Alex
2     John
3     Emma
4    David
Name: name, dtype: object

Selected columns:
    name    city
0   Sara   Dhaka
1   Alex  London
2   John   Paris
3   Emma   Tokyo
4  David  Sydney

Selected rows and columns:
   name  age    city
1  Alex   30  London
2  John   22   Paris
3  Emma   28   Tokyo

Specific value:
London

Non-consecutive rows:
    name  age    city  score
0   Sara   25   Dhaka     85
2   John   22   Paris     78
4  David   35  Sydn

### Pandas `DataFrame.drop_duplicates()`

>`DataFrame.drop_duplicates()` removes duplicate rows from a DataFrame.

#### Example

> Remove completely duplicate rows
```python
all_duplicates_removed = duplicate_df.drop_duplicates()
```
> Remove duplicates based on the "name" column
```python
name_duplicates_removed = duplicate_df.drop_duplicates(
    subset=["name"]
)
```
> Keep the last duplicate occurrence
```python
keep_last_duplicates = duplicate_df.drop_duplicates(
    subset=["name"],
    keep="last"
)
```
> Remove all rows containing duplicate names
```python
remove_all_duplicates = duplicate_df.drop_duplicates(
    subset=["name"],
    keep=False
)
```
> Remove duplicates and reset the index
```python
reset_index_df = duplicate_df.drop_duplicates(
    ignore_index=True
)
```
> Update the original DataFrame

```python 
duplicate_df.drop_duplicates(
    inplace=True,
    ignore_index=True
)
```

## Basic syntax

```python
duplicate_df.drop_duplicates(
    subset=None,
    keep="first",
    inplace=False,
    ignore_index=False
)
```

## Important parameters

* `subset`: Columns used to identify duplicate rows.
* `keep="first"`: Keeps the first duplicate occurrence.
* `keep="last"`: Keeps the last duplicate occurrence.
* `keep=False`: Removes every duplicated occurrence.
* `inplace=True`: Updates the original DataFrame.
* `ignore_index=True`: Resets the index after removing duplicates.

## Common examples

```python
duplicate_df.drop_duplicates()

duplicate_df.drop_duplicates(subset=["name"])

duplicate_df.drop_duplicates(subset=["name", "city"])

duplicate_df.drop_duplicates(subset=["name"], keep="last")

duplicate_df.drop_duplicates(subset=["name"], keep=False)

duplicate_df.drop_duplicates(ignore_index=True)

duplicate_df.drop_duplicates(inplace=True)
```


In [44]:

duplicate_df = pd.DataFrame({
    "name": ["Sara", "Alex", "Sara", "John", "Alex", "Emma"],
    "age": [25, 30, 25, 22, 30, 28],
    "city": ["Dhaka", "London", "Dhaka", "Paris", "London", "Dhaka"]
})

print("Original DataFrame:")
print(duplicate_df)

# Remove completely duplicate rows
all_duplicates_removed = duplicate_df.drop_duplicates()

print("\nAfter removing completely duplicate rows:")
print(all_duplicates_removed)

# Remove duplicates based only on the "name" column
name_duplicates_removed = duplicate_df.drop_duplicates(
    subset=["name"]
)

print("\nAfter removing duplicate names:")
print(name_duplicates_removed)

# Keep the last duplicate occurrence
keep_last_duplicates = duplicate_df.drop_duplicates(
    subset=["name"],
    keep="last"
)

print("\nKeep the last occurrence:")
print(keep_last_duplicates)

# Remove every row that has a duplicate name
remove_all_duplicates = duplicate_df.drop_duplicates(
    subset=["name"],
    keep=False
)

print("\nRemove all duplicated names:")
print(remove_all_duplicates)

# Remove duplicates and reset the index
reset_index_df = duplicate_df.drop_duplicates(
    ignore_index=True
)

print("\nRemove duplicates and reset index:")
print(reset_index_df)

# Modify the original DataFrame directly
duplicate_df.drop_duplicates(
    inplace=True,
    ignore_index=True
)

print("\nUpdated original DataFrame:")
print(duplicate_df)

Original DataFrame:
   name  age    city
0  Sara   25   Dhaka
1  Alex   30  London
2  Sara   25   Dhaka
3  John   22   Paris
4  Alex   30  London
5  Emma   28   Dhaka

After removing completely duplicate rows:
   name  age    city
0  Sara   25   Dhaka
1  Alex   30  London
3  John   22   Paris
5  Emma   28   Dhaka

After removing duplicate names:
   name  age    city
0  Sara   25   Dhaka
1  Alex   30  London
3  John   22   Paris
5  Emma   28   Dhaka

Keep the last occurrence:
   name  age    city
2  Sara   25   Dhaka
3  John   22   Paris
4  Alex   30  London
5  Emma   28   Dhaka

Remove all duplicated names:
   name  age   city
3  John   22  Paris
5  Emma   28  Dhaka

Remove duplicates and reset index:
   name  age    city
0  Sara   25   Dhaka
1  Alex   30  London
2  John   22   Paris
3  Emma   28   Dhaka

Updated original DataFrame:
   name  age    city
0  Sara   25   Dhaka
1  Alex   30  London
2  John   22   Paris
3  Emma   28   Dhaka


## Aggregation

### Explore DataFrame value_counts()

In [55]:
print(type(df))
print(df['PaymentMethod'].value_counts())

<class 'pandas.core.frame.DataFrame'>
PaymentMethod
Card              111
Bank Transfer     108
Cash              106
Mobile Banking    105
MFS                 5
cash                5
CASH                5
CARD                4
card                4
BankTransfer        3
MobileBanking       3
bank transfer       1
Name: count, dtype: int64


### Explore DataFrame null counts

In [59]:
print(df.isnull().sum())

OrderID                0
OrderDate              0
CustomerID             0
CustomerName          22
CustomerAge            9
Email                  1
Region                 0
City                  10
Category               0
Product                0
Quantity               0
UnitPrice             13
DiscountPct           17
PaymentMethod          0
SalesRep               0
Returned               0
SatisfactionScore     50
DeliveryDays          11
Notes                209
dtype: int64


### Pandas `DataFrame.groupby()`

>Brief definition

`DataFrame.groupby()` groups rows based on one or more columns and applies calculations to each group.

> Correct method name: `groupby()`

**Basic syntax**

```python
groupby_df.groupby("column_name")
```

**Example DataFrame**

```python
import pandas as pd

groupby_df = pd.DataFrame({
    "department": ["IT", "IT", "HR", "HR", "Sales", "Sales"],
    "employee": ["Sara", "Alex", "John", "Emma", "David", "Sophia"],
    "city": ["Dhaka", "London", "Dhaka", "London", "Dhaka", "London"],
    "salary": [50000, 60000, 45000, 48000, 55000, 65000],
    "sales": [20, 25, 10, 15, 30, 35]
})
```

**Total salary by department**

```python
groupby_df.groupby("department")["salary"].sum()
```

**Average salary by department**

```python
groupby_df.groupby("department")["salary"].mean()
```

**Highest salary by department**

```python
groupby_df.groupby("department")["salary"].max()
```

**Lowest salary by department**

```python
groupby_df.groupby("department")["salary"].min()
```

**Number of rows in each group**

```python
groupby_df.groupby("department").size()
```

**Count non-missing values**

```python
groupby_df.groupby("department")["employee"].count()
```

**Group by multiple columns**

```python
groupby_df.groupby(["city", "department"])["salary"].sum()
```

**Keep the grouped column as a normal column**

```python
groupby_df.groupby(
    "department",
    as_index=False
)["salary"].sum()
```

**Apply multiple calculations**

```python
groupby_df.groupby("department")["salary"].agg([
    "sum",
    "mean",
    "min",
    "max"
])
```

**Named aggregation**

```python
groupby_df.groupby("department").agg(
    total_salary=("salary", "sum"),
    average_salary=("salary", "mean"),
    highest_salary=("salary", "max")
)
```

**Reset the index**

```python
groupby_df.groupby("department")["salary"].sum().reset_index()
```

**Add group average to every row**

```python
groupby_df["average_salary"] = (
    groupby_df
    .groupby("department")["salary"]
    .transform("mean")
)
```

**Filter groups**

```python
groupby_df.groupby("department").filter(
    lambda group: group["sales"].sum() > 40
)
```

$$Common aggregation methods$$
#### Common `groupby().agg()` methods

`first`

Returns the first non-null value from each group.

`last`

Returns the last non-null value from each group.

`sum`

Returns the total of all values in each group.

`mean`

Returns the average value of each group.

`median`

Returns the median value of each group.

`min`

Returns the minimum value from each group.

`max`

Returns the maximum value from each group.

`count`

Returns the number of non-null values in each group.

`size`

Returns the total number of rows in each group, including null values.

`nunique`

Returns the number of unique values in each group.

`prod`

Returns the product of all values in each group.

`std`

Returns the standard deviation of each group.

`var`

Returns the variance of each group.

`sem`

Returns the standard error of the mean for each group.

`quantile`

Returns a specified quantile value from each group.

`any`

Returns `True` if at least one value in the group is true.

`all`

Returns `True` if all values in the group are true.

`idxmin`

Returns the index of the minimum value in each group.

`idxmax`

Returns the index of the maximum value in each group.

`skew`

Returns the skewness of values in each group.

`cov`

Returns the covariance between columns within each group.


In [60]:

groupby_df = pd.DataFrame({
    "department": ["IT", "IT", "HR", "HR", "Sales", "Sales"],
    "employee": ["Sara", "Alex", "John", "Emma", "David", "Sophia"],
    "city": ["Dhaka", "London", "Dhaka", "London", "Dhaka", "London"],
    "salary": [50000, 60000, 45000, 48000, 55000, 65000],
    "sales": [20, 25, 10, 15, 30, 35]
})

print("Original DataFrame:")
print(groupby_df)

# Total salary for each department
department_salary = groupby_df.groupby("department")["salary"].sum()

print("\nTotal salary by department:")
print(department_salary)

# Average salary for each department
average_salary = groupby_df.groupby("department")["salary"].mean()

print("\nAverage salary by department:")
print(average_salary)

# Number of employees in each department
employee_count = groupby_df.groupby("department").size()

print("\nEmployee count by department:")
print(employee_count)

# Group using multiple columns
city_department_salary = (
    groupby_df
    .groupby(["city", "department"])["salary"]
    .sum()
)

print("\nSalary grouped by city and department:")
print(city_department_salary)

# Apply multiple calculations
department_summary = (
    groupby_df
    .groupby("department")
    .agg(
        total_salary=("salary", "sum"),
        average_salary=("salary", "mean"),
        highest_salary=("salary", "max"),
        total_sales=("sales", "sum")
    )
    .reset_index()
)

print("\nDepartment summary:")
print(department_summary)

# Keep department as a normal column
grouped_with_column = (
    groupby_df
    .groupby("department", as_index=False)["salary"]
    .sum()
)

print("\nGrouped result with a normal index:")
print(grouped_with_column)

# Add department average beside every employee
groupby_df["department_average_salary"] = (
    groupby_df
    .groupby("department")["salary"]
    .transform("mean")
)

print("\nDataFrame with department average salary:")
print(groupby_df)

# Keep groups whose total sales are greater than 40
filtered_groups = groupby_df.groupby("department").filter(
    lambda group: group["sales"].sum() > 40
)

print("\nDepartments with total sales greater than 40:")
print(filtered_groups)

Original DataFrame:
  department employee    city  salary  sales
0         IT     Sara   Dhaka   50000     20
1         IT     Alex  London   60000     25
2         HR     John   Dhaka   45000     10
3         HR     Emma  London   48000     15
4      Sales    David   Dhaka   55000     30
5      Sales   Sophia  London   65000     35

Total salary by department:
department
HR        93000
IT       110000
Sales    120000
Name: salary, dtype: int64

Average salary by department:
department
HR       46500.0
IT       55000.0
Sales    60000.0
Name: salary, dtype: float64

Employee count by department:
department
HR       2
IT       2
Sales    2
dtype: int64

Salary grouped by city and department:
city    department
Dhaka   HR            45000
        IT            50000
        Sales         55000
London  HR            48000
        IT            60000
        Sales         65000
Name: salary, dtype: int64

Department summary:
  department  total_salary  average_salary  highest_salary  total

### Pandas `DataFrame.sort_values()`

> Brief definition

>`DataFrame.sort_values()` sorts DataFrame rows by the values of one or more columns.

**Basic syntax**

```python
sort_values_df.sort_values(
    by="column_name",
    ascending=True
)
```

**Sort one column in ascending order**

```python
sort_values_df.sort_values(by="price")
```

**Sort one column in descending order**

```python
sort_values_df.sort_values(
    by="price",
    ascending=False
)
```

**Sort using multiple columns**

```python
sort_values_df.sort_values(
    by=["category", "price"]
)
```

**Use different sorting orders**

```python
sort_values_df.sort_values(
    by=["category", "price"],
    ascending=[True, False]
)
```

**Place missing values first**

```python
sort_values_df.sort_values(
    by="price",
    na_position="first"
)
```

**Place missing values last**

```python
sort_values_df.sort_values(
    by="price",
    na_position="last"
)
```

**Reset the index after sorting**

```python
sort_values_df.sort_values(
    by="price",
    ignore_index=True
)
```

**Update the original DataFrame**

```python
sort_values_df.sort_values(
    by="price",
    inplace=True
)
```

### Important parameters

**`by`**

>Specifies the column or columns used for sorting.

**`ascending=True`**

Sorts values from smallest to largest.

**`ascending=False`**

Sorts values from largest to smallest.

**`na_position="first"`**

Places missing values at the beginning.

**`na_position="last"`**

Places missing values at the end.

**`ignore_index=True`**

Creates a new sequential index after sorting.

**`inplace=True`**

Updates the original DataFrame directly.


In [61]:

sort_values_df = pd.DataFrame({
    "item": ["Laptop", "Mouse", "Keyboard", "Monitor", "Headphone"],
    "category": [
        "Computer",
        "Accessory",
        "Accessory",
        "Computer",
        "Accessory"
    ],
    "price": [800, 20, 50, 300, 80],
    "quantity": [5, 20, 10, 7, 12]
})

print("Original DataFrame:")
print(sort_values_df)

# Sort price in ascending order
price_ascending = sort_values_df.sort_values(
    by="price"
)

print("\nPrice in ascending order:")
print(price_ascending)

# Sort price in descending order
price_descending = sort_values_df.sort_values(
    by="price",
    ascending=False
)

print("\nPrice in descending order:")
print(price_descending)

# Sort using category and price
multiple_column_sort = sort_values_df.sort_values(
    by=["category", "price"],
    ascending=[True, False]
)

print("\nSort by category ascending and price descending:")
print(multiple_column_sort)

# Sort quantity and reset the index
quantity_sorted = sort_values_df.sort_values(
    by="quantity",
    ascending=False,
    ignore_index=True
)

print("\nQuantity sorted with a new index:")
print(quantity_sorted)

# Update the original DataFrame
sort_values_df.sort_values(
    by="price",
    ascending=True,
    ignore_index=True,
    inplace=True
)

print("\nUpdated original DataFrame:")
print(sort_values_df)

Original DataFrame:
        item   category  price  quantity
0     Laptop   Computer    800         5
1      Mouse  Accessory     20        20
2   Keyboard  Accessory     50        10
3    Monitor   Computer    300         7
4  Headphone  Accessory     80        12

Price in ascending order:
        item   category  price  quantity
1      Mouse  Accessory     20        20
2   Keyboard  Accessory     50        10
4  Headphone  Accessory     80        12
3    Monitor   Computer    300         7
0     Laptop   Computer    800         5

Price in descending order:
        item   category  price  quantity
0     Laptop   Computer    800         5
3    Monitor   Computer    300         7
4  Headphone  Accessory     80        12
2   Keyboard  Accessory     50        10
1      Mouse  Accessory     20        20

Sort by category ascending and price descending:
        item   category  price  quantity
4  Headphone  Accessory     80        12
2   Keyboard  Accessory     50        10
1      Mouse  

### Pandas `DataFrame.apply()`

>Brief definition

`DataFrame.apply()` applies a function to each column or each row of a DataFrame.

It does not normally apply the function to every individual cell. For element-wise operations, use `DataFrame.map()`.

**Basic syntax**

```python
apply_df.apply(function)
```

**Apply to each column**

`axis=0` applies the function to each column.

```python
apply_df.apply(function, axis=0)
```

**Apply to each row**

`axis=1` applies the function to each row.

```python
apply_df.apply(function, axis=1)
```

**Find the maximum of each numeric column**

```python
apply_df[["price", "quantity"]].apply(max)
```

**Find the sum of each numeric column**

```python
apply_df[["price", "quantity"]].apply(sum)
```

**Use a lambda function on columns**

```python
apply_df[["price", "quantity"]].apply(
    lambda column: column.mean()
)
```

## Create a value using multiple columns

```python
apply_df["total"] = apply_df.apply(
    lambda row: row["price"] * row["quantity"],
    axis=1
)
```

**Create a category using row values**

```python
apply_df["price_level"] = apply_df.apply(
    lambda row: "Expensive" if row["price"] >= 100 else "Affordable",
    axis=1
)
```

**Apply a named function**

```python
def calculate_total(row):
    return row["price"] * row["quantity"]
```

```python
apply_df["total"] = apply_df.apply(
    calculate_total,
    axis=1
)
```

**Return multiple columns**

```python
def calculate_details(row):
    return pd.Series({
        "total": row["price"] * row["quantity"],
        "available": row["quantity"] > 0
    })
```

```python
details = apply_df.apply(
    calculate_details,
    axis=1
)
```

### Important parameters

`func`

The function that will be applied.

`axis=0`

Applies the function to each column. This is the default.

`axis=1`

Applies the function to each row.

`args`

Passes additional positional arguments to the function.

`result_type="expand"`

Expands list-like row results into separate columns.

`apply()` versus `map()`

Use `apply()` when the function needs an entire row or column.

```python
apply_df.apply(function, axis=1)
```

Use `map()` when the function should process every individual cell.

```python
apply_df.map(function)
```


In [62]:


apply_df = pd.DataFrame({
    "item": ["Laptop", "Mouse", "Keyboard", "Monitor"],
    "price": [800, 20, 50, 300],
    "quantity": [2, 10, 5, 3]
})

print("Original DataFrame:")
print(apply_df)

# Apply sum to each selected column
column_totals = apply_df[["price", "quantity"]].apply(
    sum,
    axis=0
)

print("\nColumn totals:")
print(column_totals)

# Apply mean to each selected column
column_averages = apply_df[["price", "quantity"]].apply(
    lambda column: column.mean(),
    axis=0
)

print("\nColumn averages:")
print(column_averages)

# Calculate total price using each row
apply_df["total_price"] = apply_df.apply(
    lambda row: row["price"] * row["quantity"],
    axis=1
)

# Create a category using each row
apply_df["price_level"] = apply_df.apply(
    lambda row: "Expensive"
    if row["price"] >= 100
    else "Affordable",
    axis=1
)

# Apply a named function
def calculate_discount(row):
    if row["quantity"] >= 5:
        return row["total_price"] * 0.10
    return 0


apply_df["discount"] = apply_df.apply(
    calculate_discount,
    axis=1
)

# Calculate final price
apply_df["final_price"] = apply_df.apply(
    lambda row: row["total_price"] - row["discount"],
    axis=1
)

print("\nUpdated DataFrame:")
print(apply_df)

Original DataFrame:
       item  price  quantity
0    Laptop    800         2
1     Mouse     20        10
2  Keyboard     50         5
3   Monitor    300         3

Column totals:
price       1170
quantity      20
dtype: int64

Column averages:
price       292.5
quantity      5.0
dtype: float64

Updated DataFrame:
       item  price  quantity  total_price price_level  discount  final_price
0    Laptop    800         2         1600   Expensive       0.0       1600.0
1     Mouse     20        10          200  Affordable      20.0        180.0
2  Keyboard     50         5          250  Affordable      25.0        225.0
3   Monitor    300         3          900   Expensive       0.0        900.0
